res/controllers/Denon-MC4000-scripts.js

```diff
@@ -0,0 +1,544 @@
+
+    var that = this; // FIXME for 2.1: hacks around https://bugs.launchpad.net/mixxx/+bug/1567203
+
+    this.input = function (channel, control, value, status, group) {
+        if (value > 0) {
+            if (engine.getValue(that.group, 'play') === 1) {
+                engine.setValue(that.group, 'play', 0);
+            } else {
+                engine.setValue(that.group, 'eject', 1);
+            }
+        }
+    };
+}
+SamplerStopButton.prototype = Object.create(SamplerButton);
+SamplerStopButton.prototype.constructor = SamplerStopButton;
+
+
+
+var MC4000 = {};
+
+MC4000.init = function () {
+    // Decks
+    MC4000.leftDeck = new MC4000.Deck(0);
+    MC4000.rightDeck = new MC4000.Deck(1);
+    // VU meters
+    // TODO: implement VU meters for the PFL when cue button is on
+    engine.connectControl("[Master]", "VuMeterL", "MC4000.OnVuMeterChangeL");
+    engine.connectControl("[Master]", "VuMeterR", "MC4000.OnVuMeterChangeR");
+    // Samplers
+    for (var i = 1; i <= 4; i++) {
+        this['sampler' + i] = new SamplerButton([0x9F, 0x20 + i], i);
+        this['samplerClear' + i] = new SamplerStopButton([0x9F, 0x27 + i], i);
+    }
+    MC4000.samplerLevel = function (channel, control, value, status, group) {
+        engine.setValue("[Sampler1]", "pregain", script.absoluteNonLin(value, 0, 1.0, 4.0));
+        engine.setValue("[Sampler2]", "pregain", script.absoluteNonLin(value, 0, 1.0, 4.0));
+        engine.setValue("[Sampler3]", "pregain", script.absoluteNonLin(value, 0, 1.0, 4.0));
+        engine.setValue("[Sampler4]", "pregain", script.absoluteNonLin(value, 0, 1.0, 4.0));
+    }
+    // Initalization of controller states
+    // TODO: find equivalent message for MC4000
+    // tell controller to send MIDI messages with positions of faders and knobs
+    // midi.sendShortMsg(0xB0, 0x7F, 0x7F);
+};
+
+MC4000.shutdown = function () {};
+
+MC4000.padColors = {
+    off: 0,
+    red: 2,
+    blue: 1,
+    purple: 3
+};
+
+// Found by averaging total 'ticks' received over 100 rotations
+MC4000.jogWheelTicksPerRevolution = 605;
+
+MC4000.scratchParams = {
+    recordSpeed: 33.3333333333,
+    alpha: (1.0/8),
+    beta: (1.0/8)/32.0
+};
+
+MC4000.jogParams = {
+    sensitivity: 25,
+    maxJogValue: 3
+};
+
+MC4000.rateRanges = [0.08, 0.16, 0.5];
+
+MC4000.OnVuMeterChangeL = function(value, group, control) {
+    midi.sendShortMsg(0xBF, 0x44, value*0x7F);
+}
+
+MC4000.OnVuMeterChangeR = function(value, group, control) {
+    midi.sendShortMsg(0xBF, 0x45, value*0x7F);
+}
+
+MC4000.Deck = function (channel) {
+    // Some state variables
+    this.isVinylMode = false;
+    this.rateMsbValue = 0;
+    this.midiChannel = channel;
+    this.engineChannel = channel + 1;
+    this.group = '[Channel' + this.engineChannel + ']';
+    this.currentRateIndex = 0;
+    
+    // Set default rate range (to +/- 8% like Serato and Technics)
+    engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
```

Comment:
Alright, I guess I could hardcode the options listed in preferences into
the JS and toggle through that.
On 7 May 2016 10:59, "RJ Ryan" <notifications@github.com> wrote:

> In res/controllers/Denon-MC4000-scripts.js
> <https://github.com/mixxxdj/mixxx/pull/938#discussion_r62409476>:
>
> > +
> > +MC4000.OnVuMeterChangeR = function(value, group, control) {
> > +    midi.sendShortMsg(0xBF, 0x45, value*0x7F);
> > +}
> > +
> > +MC4000.Deck = function (channel) {
> > +    // Some state variables
> > +    this.isVinylMode = false;
> > +    this.rateMsbValue = 0;
> > +    this.midiChannel = channel;
> > +    this.engineChannel = channel + 1;
> > +    this.group = '[Channel' + this.engineChannel + ']';
> > +    this.currentRateIndex = 0;
> > +
> > +    // Set default rate range (to +/- 8% like Serato and Technics)
> > +    engine.setValue(this.group, "rateRange", MC4000.rateRanges[this.currentRateIndex]);
>
> @Be-ing <https://github.com/Be-ing> is right here, you can access the
> user's preferred rate range via the "rateRange" control.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/938/files/4dcee7f172408d704ffea276057901605bb8d7b5#r62409476>
>